oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java

```diff
@@ -92,8 +100,7 @@ public Authentication authenticate(Authentication authentication) throws Authent
 		}
 		BearerTokenAuthenticationToken bearer = (BearerTokenAuthenticationToken) authentication;
 		OAuth2AuthenticatedPrincipal principal = getOAuth2AuthenticatedPrincipal(bearer);
-		AbstractAuthenticationToken result = convert(principal, bearer.getToken());
-		result.setDetails(bearer.getDetails());
```

Comment:
One thing I just noticed, we should ensure the details are still being set if applicable. I believe it would be acceptable to introduce an check for `AbstractAuthenticationToken` and cast/set the details if true. Similar to:

https://github.com/spring-projects/spring-security/blob/d996c2a2c6d14c4e578cb56019737922060fd7da/saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java#L138-L143

I'm currently preferring that instead of returning `AbstractAuthenticationToken` from our converter.

**Note:** We don't need to introduce a `AuthenticationDetailsSource` at this time.

Reply:
I'm not found of the idea to alter the Authentication instance after the authenticationConverter returned it. Maybe limit that to just `BearerAuthenticationToken` instances instead of all `AbstractAuthenticationToken`?

Reply:
I think I found a way to do it from inside default converter. Please let me know what you think of my latest commit.

Reply:
Thanks @ch4mpy, it's an interesting idea! However, I'm not so sure in this case. As far as I can tell, this solution only works when the default converter is used, and introduces more code to do so. Doing a quick search for `setDetails` in the code base, it's fairly common for a filter or provider to set the details afterwards. I think adding a check and using `AbstractAuthenticationToken` provides good flexibility and consistency with the rest of the framework.

Reply:
> this solution only works when the default converter is used

This is exactly my intention. Concider this sample (it compiles)
```java
	public static class KeycloakClaims extends HashMap<String, Object> implements Serializable, StandardClaimAccessor {
		@Override
		public Map<String, Object> getClaims() {
			return this;
		}
		
		public List<String> getRoles() {
			final var realmAccess = (Map<String, Object>) getOrDefault("realm_access", Map.of());
			return (List<String>) realmAccess.getOrDefault("roles", List.of());
		}
	}
	
	public static class RealmAuthentication extends AbstractAuthenticationToken {

		public RealmAuthentication(KeycloakClaims claims) {
			super(claims.getRoles().stream().map(SimpleGrantedAuthority::new).toList());
			this.setDetails(claims);
		}

		@Override
		public void setDetails(Object details) {
			assert(KeycloakClaims.class.isAssignableFrom(details.getClass()));
			super.setDetails(details);
		}

		@Override
		public KeycloakClaims getDetails() {
			return (KeycloakClaims)super.getDetails();
		}

		@Override
		public KeycloakClaims getCredentials() {
			return this.getDetails();
		}

		@Override
		public KeycloakClaims getPrincipal() {
			return this.getDetails();
		}
	}
```
Clearly, trying to set details with a random `Object` after the authenticationConverter returned such an `AbstractAuthenticationToken` implementation will be a problem.

The aim of this request is to give total control over `Authentication` instance to the framework user after successful introspection.

I understand the need to keep exact same behaviour for default authenticationConverter, but I think it would be very tricky (and frustrating) for one providing explicitly an authenticationConverter to have the authentication instance it returned be arbitrarly altered (with value out of his control).

## PS
As far I can read from existing code, `details` are set for servlets only. Is it intended or just a bug introduced when duplicating code?

Reply:
Those are good points, @ch4mpy.

> The aim of this pull request is to give total control over `Authentication` instance to the framework user after successful introspection.

It may be that a converter is not the best solution in that case? Introducing a converter adds a configuration hook that gives flexibility to the user. If total control is the goal, there are other ways to accomplish that. For example, we could build a delegating `AuthenticationProvider`:

```java
public class MyOpaqueTokenAuthenticationProvider implements AuthenticationProvider {

	private final AuthenticationProvider delegate;

	public MyOpaqueTokenAuthenticationProvider(OpaqueTokenIntrospector introspector) {
		this.delegate = new OpaqueTokenAuthenticationProvider(introspector);
	}

	@Override
	public boolean supports(Class<?> authentication) {
		return this.delegate.supports(authentication);
	}

	@Override
	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
		Authentication result = this.delegate.authenticate(authentication);
		// ...
	}

}
```

This gives full control at the expense of constructing a new authentication based on the returned one. Of course, you can always choose to fully implement the `AuthenticationProvider` yourself if needed.

Having said that, it sounds like my suggestion conflicts with your intended use case for introducing the converter? If that's the case, perhaps your authentication instance shouldn't extend from `AbstractAuthenticationToken`? Or it could simply store the claims in another field, which seems to make sense here. The claims in your example seem more specific than `Object details` from `AbstractAuthenticationToken`.

> As far I can read from existing code, `details` are set for servlets only. Is it intended or just a bug introduced when duplicating code in `OpaqueTokenReactiveAuthenticationManager`?

I'd have to dig into it to say for 100% certain, but it could be that the use of `details` addresses a servlet-specific need.

Reply:
> The claims in your example seem more specific than `Object details` from `AbstractAuthenticationToken`.

Hopefully, yes, it is. To be honest, I like typed programming languages and wish there were no `Object` return type in the framework (`Authentication` and `AbstractAuthenticationToken` could use generics or something), but this is not the subject here.

> perhaps your authentication instance shouldn't extend from AbstractAuthenticationToken?

Implementing just `Authentication` would keep such an implementation away of all the part of the framework which expect `AbstractAuthenticationToken`?

> could simply store the claims in another field

Yes, maybe. I just like the idea of specializations (`B extends A`) to be more specific, specially when `A` is way too generic (like having an `Object` return type)

If you really want to keep the side effect on details for all `AbstractAuthenticationToken`, inside servlets only, then maybe should it be carefully documented. The `Authentication` implementations I write can live with it and my security expressions will just omit `details` property (if I don't know what is in there, I will certainely not use it).

Are you sure you don't want the test to apply the `details` side effect to be applied on `OpaqueTokenAuthentication` instances only instead of all `AbstractAuthenticationToken`?

Reply:
Thanks for explaining your point of view, @ch4mpy. I think if we were designing a new component entirely (new authentication provider or filter) we would have quite a bit more license to use these ideas to define it ourselves.

> Are you sure you don't want the `details` side effect to be applied to `BearerTokenAuthentication` instances only (instead of all `AbstractAuthenticationToken`)?

This is not what the `AuthenticationProvider` does today, so technically this would be a breaking change to the default behavior, which is mainly what I'm proposing we try and avoid.

It's worth noting that once this authentication is returned to the framework, the `ProviderManager` does exactly what I'm proposing but with a `null` check:

https://github.com/spring-projects/spring-security/blob/929b3345886c1810b32cd38257bf19c629008b92/core/src/main/java/org/springframework/security/authentication/ProviderManager.java#L258-L263

If you want to add the same `null` check and not overwrite what was set in the converter, I think that would be a fair compromise and wouldn't be a breaking change. What do you think?

Reply:
Yes, I prefer to add a null check to set details only if was "forgotten".

Should this be applied to reactive apps too?